### ISSUE TITLE
Pointer ID strings are currently not really cross platform

### ISSUE THREAD

**Issue by Spacechild1** [pure-data/pure-data#1474]
There are hundreds of places in the Pd core where we generate a string from a memory address. This is generally done using the `%lx` specifier.

The problem is, passing a pointer to `%lx` is only correct on LP64 platforms. On LLP64 platforms, like Windows, `long` is always 32-bit, so we effectively only use the lower 32 bits of the address. In practice this is fine most of the time because it is sufficiently unique, but there can be situations where two pointers have the same lower 32 bits and are only distinguished by the upper 32 bits. In that case, both objects would have the same ID and the Pd GUI would get confused.

I think the proper solution is to get rid of the explicit `%lx` in the code base and use some function or macro instead. For LP64 platforms it can resolve to `%lx`. On Windows we could use `%Ix`.

This could be a simple macro:

```
#ifdef _WIN32
# define PTR_ID "%Ix"
#else
# define PTR_ID "%lx"
#endif
```

Additionally, we could have macros for different kinds of IDs:
```
#define OBJ_ID ".x" PTR_ID
#define CANVAS_ID OBJ_ID ".c"
#define LINE_ID "l" PTR_ID
// etc.
```

Usage:
```
    sys_vgui(
        CANVAS_ID " create line %d %d %d %d -width %d -tags [list " LINE_ID " cord]\n",
                glist_getcanvas(x),
                t.tr_lx1, t.tr_ly1, t.tr_lx2, t.tr_ly2,
                (outlet_getsymbol(t.tr_outlet) == &s_signal ? 2:1) * x->gl_zoom,
                oc);
```

---

Alternative, we could also have functions that write the IDs into a string buffer, e.g.
```
static char* ptr_to_id(const void *pointer, char *dest, size_t len)
{
    *dest=0;
    if (pointer)
#ifdef _WIN32
        snprintf(dest, len, "%Ix", pointer);
#else
        snprintf(dest, len, "%lx", pointer);
#endif
    return dest;
}

static char* obj_to_id(const void *pointer, char *dest, size_t len)
{
    *dest=0;
    if (pointer)
#ifdef _WIN32
        snprintf(dest, len, ".x%Ix", pointer);
#else
        snprintf(dest, len, ".x%lx", pointer);
#endif
    return dest;
}

static char* canvas_to_id(const void *pointer, char *dest, size_t len)
{
    *dest=0;
    if (pointer)
#ifdef _WIN32
        snprintf(dest, len, ".x%Ix.c", pointer);
#else
        snprintf(dest, len, ".x%lx.c", pointer);
#endif
    return dest;
}

static char* line_to_id(const void *pointer, char *dest, size_t len)
{
    *dest=0;
    if (pointer)
#ifdef _WIN32
        snprintf(dest, len, "l%Ix", pointer);
#else
        snprintf(dest, len, "l%lx", pointer);
#endif
    return dest;
}

// similar functions for objects, canvases, lines, etc.
```

Usage:

```
    char buf1[64], buf2[64];
    sys_vgui(
        "%s create line %d %d %d %d -width %d -tags [list %s cord]\n",
                canvas_to_id(glist_getcanvas(x), buf1, sizeof(buf1)),
                t.tr_lx1, t.tr_ly1, t.tr_lx2, t.tr_ly2,
                (outlet_getsymbol(t.tr_outlet) == &s_signal ? 2:1) * x->gl_zoom,
                line_to_id(oc, buf2, sizeof(buf2)));
```

---

Personally, I tend to prefer the macro solution. It's less code, has no performance overhead, easier to read and the specifiers stand out visually.

**Comment by umlaeute** [comment]
i tend to prefer wrapper functions as they simply allow us to do more things when mangling the ID. 
also i find them actually more readable than strings-intermangled-with-macros which make me eye-sore (though i get your point of having a visual marker in a longish format string).

performance might be an issue though.

**Comment by Spacechild1** [comment]
> strings-intermangled-with-macros

As a side note: the cross-platform format specifier macros in `<inttypes.h>`, like `PRId64`, are used the same way. Of course, you would find those probably ugly as well. I just wanted to point out that the macro solution wouldn't be something unusual.

**Comment by umlaeute** [comment]
it's not a matter of the actual name of the macro...(nor that `define`s for format-specifiers are something unusual, they are not).

i guess it is more that what we (I) want is not to "print a pointer", but to "create an ID for an 'object'" (and it's pure coincidence that we currently use the object's memory address to generate this ID).

i have a vague memory that i once had an idea for something where the current scheme would make problems. cannot remember though :-)

**Comment by Spacechild1** [comment]
> i guess it is more that what we (I) want is not to "print a pointer", but to "create an ID for an 'object'" (and it's pure coincidence that we currently use the object's memory address to generate this ID).

Fair enough. There would be other (maybe better ways) to generate such IDs.

> i have a vague memory that i once had an idea for something where the current scheme would make problems. cannot remember though :-)

If you create an object A, delete it and create object B, the latter might end up at the same memory address as A, so they both would have the same ID. This could lead to confusion if the GUI sends a message to A while it is being replaced with B. Not very likely, but certainly possible.

A much safer technique would have been to generate IDs from a monotonically increasing counter and store them in `t_gobj`. I don't think it's possible to retrofit something like this without breaking binary compatibility...

So I am wondering if there are any other approaches that *would* be possible...

**Comment by Spacechild1** [comment]
On the other hand, why don't we just use `%p` to print the pointer? The resulting string will look different with every compiler but is this a problem?

**Comment by umlaeute** [comment]
the IDs are used as names for Tk-widgets, so they must be valid Tcl variable names.

if your compiler decides to print a pointer as `[0000:ffff]` we got a problem.

**Comment by Spacechild1** [comment]
Thanks! I knew there must be a catch... it was just too simple :-)

**Comment by umlaeute** [comment]
is there some list on how various compilers/libcs convert `%p`?

**Comment by Spacechild1** [comment]
I couldn't find one.

I quickly tried with godbolt: https://godbolt.org/z/avPzrdovf

x86_64 clang and GCC would both print something like `0x11c82a0`. (The host system is Linux, I guess.)

On the Windows, the CRT would print something like `00625be0` for 32-bit pointers and `0000000000309030` for 64-bit pointers. The leading zeros are annoying and not exactly efficient...

**Comment by aaaaalbert** [comment]
I found this: https://stackoverflow.com/a/9053835

It includes some explanation and cites from the C99 standard that `%p` is generally implementation specific (which, of course, doesn't say much about the *actual* differences between implementations).

The post also shows an example/emulation of custom styling of a `%p`-like format specifier, summing up to `#include <inttypes.h>` and using its `unintptr_t` to cast the pointer to, combined with special macros such as `PRIXPTR` ([others exits](https://pubs.opengroup.org/onlinepubs/9699919799/)) for the concrete formatting.

(Disclaimer: This is only desk research, I haven't tried this out.)

**Comment by giuliomoro** [comment]
What about something like:

```
#include <stdio.h>
#include <inttypes.h>

#if defined(__x86_64__) || defined(__amd64__) || defined(_M_X64) || defined(_M_AMD64) || defined(__aarch64__)
#define PTR_SPECIFIER PRIu64
#else
#define PTR_SPECIFIER PRIu32
#endif

int main()
{
    void* myptr = (void*)12345;
    printf("%" PTR_SPECIFIER "\n", (size_t)myptr);
    return 0;
}
```

**Comment by Spacechild1** [comment]
@giuliomoro How is that different to what I've suggested in the PR description? ;-)

**Comment by giuliomoro** [comment]
> How is that different to what I've suggested in the PR description? ;-)

uops sorry, didn't read carefully (or at all)!

**Comment by Spacechild1** [comment]
no worries!

**Comment by Spacechild1** [comment]
@umlaeute what do you think about 3? ;-)
